x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java

```diff
@@ -0,0 +1,158 @@
+        super(MountSearchableSnapshotAction.NAME, transportService, clusterService, threadPool, actionFilters,
+            MountSearchableSnapshotRequest::new, indexNameExpressionResolver);
+        this.client = client;
+        this.repositoriesService = repositoriesService;
+    }
+
+    @Override
+    protected String executor() {
+        // Using the generic instead of the snapshot threadpool here as the snapshot threadpool might be blocked on long running tasks
+        // which would block the request from getting an error response because of the ongoing task
+        return ThreadPool.Names.GENERIC;
+    }
+
+    @Override
+    protected RestoreSnapshotResponse read(StreamInput in) throws IOException {
+        return new RestoreSnapshotResponse(in);
+    }
+
+    @Override
+    protected ClusterBlockException checkBlock(MountSearchableSnapshotRequest request, ClusterState state) {
+        // Restoring a snapshot might change the global state and create/change an index,
+        // so we need to check for METADATA_WRITE and WRITE blocks
+        ClusterBlockException blockException = state.blocks().globalBlockedException(ClusterBlockLevel.METADATA_WRITE);
+        if (blockException != null) {
+            return blockException;
+        }
+        return state.blocks().globalBlockedException(ClusterBlockLevel.WRITE);
+
+    }
+
+    /**
+     * Return the index settings required to make a snapshot searchable
+     */
+    private static Settings getIndexSettings(String repoName, SnapshotId snapshotId, IndexId indexId) {
+        return Settings.builder()
+            .put(SearchableSnapshotRepository.SNAPSHOT_REPOSITORY_SETTING.getKey(), repoName)
+            .put(SearchableSnapshotRepository.SNAPSHOT_SNAPSHOT_NAME_SETTING.getKey(), snapshotId.getName())
+            .put(SearchableSnapshotRepository.SNAPSHOT_SNAPSHOT_ID_SETTING.getKey(), snapshotId.getUUID())
+            .put(SearchableSnapshotRepository.SNAPSHOT_INDEX_ID_SETTING.getKey(), indexId.getId())
+            .put(INDEX_STORE_TYPE_SETTING.getKey(), SearchableSnapshotRepository.SNAPSHOT_DIRECTORY_FACTORY_KEY)
+            .put(IndexMetaData.SETTING_BLOCKS_WRITE, true)
+            .build();
+    }
+
+    @Override
+    protected void masterOperation(Task task, final MountSearchableSnapshotRequest request, final ClusterState state,
+                                   final ActionListener<RestoreSnapshotResponse> listener) {
+        final String repoName = request.repository();
+        final String snapName = request.snapshot();
+        final String indexName = request.snapshotIndex();
+
+        // Retrieve IndexId and SnapshotId instances, which are then used to create a new restore
+        // request, which is then sent on to the actual snapshot restore mechanism
+        final Repository repository = repositoriesService.repository(repoName);
```

Comment:
This throws a `RepositoryMissingException` which must be caught and passed to the `listener.onFailure()` method

Reply:
I think throwing is fine, we're inside an `ActionRunnable`?

Reply:
Good point @DaveCTurner, but I still think that mixing listener and throwing exception is confusing

Reply:
I prefer to throw. The `try/catch` is unnecessary and meant I felt compelled to investigate whether the framework was missing something.

Reply:
Okay. I've checked other places and it seems it also throws in some cases.